File: Python/day10/main.py
from dataclasses import dataclass
from enum import Enum, auto


class Direction(Enum):
    Up = auto()
    Down = auto()
    Left = auto()
    Right = auto()

    def get_opposite(self):
        if self == Direction.Up:
            return Direction.Down
        elif self == Direction.Down:
            return Direction.Up
        elif self == Direction.Left:
            return Direction.Right
        elif self == Direction.Right:
            return Direction.Left


@dataclass
class Point:
    x: int
    y: int

    def get_next_in_direction(self, direction: Direction):
        if direction == Direction.Up:
            return Point(self.x, self.y - 1)
        elif direction == Direction.Down:
            return Point(self.x, self.y + 1)
        elif direction == Direction.Left:
            return Point(self.x - 1, self.y)
        elif direction == Direction.Right:
            return Point(self.x + 1, self.y)

    def __hash__(self):
        return hash((self.x, self.y))


SYMBOL_DIRECTIONS = {
    "|": (Direction.Up, Direction.Down),
    "-": (Direction.Left, Direction.Right),
    "L": (Direction.Right, Direction.Up),
    "J": (Direction.Left, Direction.Up),
    "7": (Direction.Left, Direction.Down),
    "F": (Direction.Right, Direction.Down)
}


class Tile:

    def __init__(self, symbol: str):
        self.symbol = symbol
        self.connection_dict = {d: False for d in Direction}
        if symbol in SYMBOL_DIRECTIONS:
            self.tile_type = "pipe"
            self.connections = SYMBOL_DIRECTIONS[symbol]
            for conn in self.connections:
                self.connection_dict[conn] = True
            self.path_dict = {self.connections[0]: self.connections[1], self.connections[1]:self.connections[0]}
        elif symbol == ".":
            self.tile_type = "ground"
        elif symbol == "S":
            self.tile_type = "start"


    def __str__(self):
        return str([int(value) for value in self.connection_dict.values()])

    def __repr__(self):
        return self.__str__()



def pad_tiles(tiles: list[list[Tile]]) -> list[list[Tile]]:
    padded_tiles = []
    width = len(tiles[0])
    padded_tiles.append([Tile(".") for _ in range(width + 2)])
    for line in tiles:
        padded_tiles.append([Tile(".")] + line + [Tile(".")])
    padded_tiles.append([Tile(".") for _ in range(width + 2)])
    return padded_tiles


def get_symbol_from_direction(direction: dict[Direction, bool]) -> str:
    code = "".join([str(int(direction[d])) for d in Direction])
    if code == "1100":
        return "|"
    elif code == "0011":
        return "-"
    elif code == "1001":
        return "L"
    elif code == "1010":
        return "J"
    elif code == "0110":
        return "7"
    elif code == "0101":
        return "F"




class Map:

    def __init__(self, tiles: list[list[Tile]]) -> None:
        self.tiles = pad_tiles(tiles)
        self.start = self.find_start()
        directions = {d: False for d in Direction}
        for d in self.get_connected_directions(self.start):
            directions[d] = True
        self[self.start] = Tile(get_symbol_from_direction(directions))

    def get_connected_directions(self, point: Point) -> list[Direction]:
        directions = []
        for d in Direction:
            neighbour = self[point.get_next_in_direction(d)]
            if neighbour.connection_dict[d.get_opposite()]:
                directions.append(d)
        return directions

    def find_start(self) -> Point:
        for y, line in enumerate(self.tiles):
            for x, tile in enumerate(line):
                if tile.tile_type == "start":
                    return Point(x, y)

    def __getitem__(self, item):
        if isinstance(item, Point):
            return self.tiles[item.y][item.x]
        return self.tiles[item]

    def __setitem__(self, key, value):
        if isinstance(key, Point) and isinstance(value, Tile):
            self.tiles[key.y][key.x] = value
        else:
            self.tiles[key] = value

    def __repr__(self):
        return "\n".join(["".join([str(tile) for tile in line]) for line in self.tiles])

    def __len__(self):
        return len(self.tiles)


def load_data(filename: str) -> Map:
    tiles = []
    with open(filename) as f:
        for line in f.readlines():
            tiles.append([Tile(c) for c in line.strip()])
    return Map(tiles)


def follow_path(start: Point, end: Point, tile_map: Map) -> set[Point]:
    current = start
    direction = tile_map[start].connections[0]
    path: set[Point] = {current}
    while True:
        current = current.get_next_in_direction(direction)
        direction = tile_map[current].path_dict[direction.get_opposite()]
        path.add(current)
        if current == end:
            break
    return path


def scan_rows(tile_map: Map, path: set[Point]) -> int:
    inside_point: set[Point] = set()
    height = len(tile_map)
    width = len(tile_map[0])
    for y in range(1, height - 1):
        is_inside = False
        x = 1
        while x < width - 1:
            point = Point(x=x, y=y)
            if point not in path:
                if is_inside:
                    inside_point.add(point)
            elif tile_map[point].symbol == "|":
                is_inside = not is_inside
            elif tile_map[point].symbol == "F":
                x += 1
                while tile_map[Point(x,y)].symbol == "-":
                    x += 1
                if tile_map[Point(x, y)].symbol == "J":
                    is_inside = not is_inside
            elif tile_map[point].symbol == "L":
                x += 1
                while tile_map[Point(x,y)].symbol == "-":
                    x += 1
                if tile_map[Point(x, y)].symbol == "7":
                    is_inside = not is_inside
            x += 1
    return len(inside_point)


def solution_part_1():
    tile_map = load_data("input.txt")
    path = follow_path(tile_map.start, tile_map.start, tile_map)
    print(len(path) // 2)


def solution_part_2():
    tile_map = load_data("input.txt")
    path = follow_path(tile_map.start, tile_map.start, tile_map)
    result = scan_rows(tile_map, path)
    print(result)


def main():
    solution_part_1()
    solution_part_2()


if __name__ == "__main__":
    main()

File: typescript/day10/main.ts
import * as fs from "fs";

enum Direction {
    Up,
    Down,
    Left,
    Right,
}

const Directions = [Direction.Up, Direction.Down, Direction.Left, Direction.Right];

function opposite(direction: Direction): Direction {
    switch (direction) {
        case Direction.Up:
            return Direction.Down;
        case Direction.Down:
            return Direction.Up;
        case Direction.Left:
            return Direction.Right;
        case Direction.Right:
            return Direction.Left;
    }
}

class Point {
    x: number
    y: number

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    equals(other: Point): boolean {
        return this.x == other.x && this.y == other.y;
    }

    string(): string {
        return `${this.x}_${this.y}`;
    }

    get_neighbour_in_direction(dir: Direction): Point {
        switch (dir) {
            case Direction.Up:
                return new Point(this.x, this.y - 1);
            case Direction.Down:
                return new Point(this.x, this.y + 1);
            case Direction.Left:
                return new Point(this.x - 1, this.y);
            case Direction.Right:
                return new Point(this.x + 1, this.y);
        }
    }
}

const SYMBOL_DIRECTIONS: Record<string, [Direction, Direction]> = {
    "|": [Direction.Up, Direction.Down],
    "-": [Direction.Left, Direction.Right],
    "L": [Direction.Right, Direction.Up],
    "J": [Direction.Left, Direction.Up],
    "7": [Direction.Left, Direction.Down],
    "F": [Direction.Right, Direction.Down]
}

const CODE_SYMBOLS: Record<string, string> = {
    "1100": "|",
    "0011": "-",
    "1001": "L",
    "1010": "J",
    "0110": "7",
    "0101": "F"
}

class Tile {
    symbol: string
    connections: Map<Direction, boolean>
    path: Map<Direction, Direction>

    constructor(symbol: string) {
        this.symbol = symbol;
        this.connections = new Map<Direction, boolean>();
        this.path = new Map<Direction, Direction>();
        if (symbol in SYMBOL_DIRECTIONS) {
            for (let dir of SYMBOL_DIRECTIONS[symbol]) {
                this.connections.set(dir, true);
            }
            this.path.set(SYMBOL_DIRECTIONS[symbol][0], SYMBOL_DIRECTIONS[symbol][1]);
            this.path.set(SYMBOL_DIRECTIONS[symbol][1], SYMBOL_DIRECTIONS[symbol][0]);
        }
    }

}

function pad_tiles(tile_map: Tile[][]): Tile[][] {
    let new_tile_map: Tile[][] = [];
    let new_tiles: Tile[] = [];
    for (let x = 0; x < tile_map[0].length + 2; x++) {
        new_tiles.push(new Tile("."));
    }
    new_tile_map.push(new_tiles);
    for (let y = 0; y < tile_map.length; y++) {
        new_tiles = [];
        new_tiles.push(new Tile("."));
        for (let x = 0; x < tile_map[y].length; x++) {
            new_tiles.push(tile_map[y][x]);
        }
        new_tiles.push(new Tile("."));
        new_tile_map.push(new_tiles);
    }
    new_tiles = [];
    for (let x = 0; x < tile_map[0].length + 2; x++) {
        new_tiles.push(new Tile("."));
    }
    new_tile_map.push(new_tiles);
    return new_tile_map;

}


function load_data(filename: string): Tile[][] {
    let tile_map: Tile[][] = [];
    let lines = fs.readFileSync(filename, "utf-8").split("\n");
    for (let line of lines) {
        let tiles: Tile[] = [];
        for (let c of line) {
            let tile = new Tile(c);
            tiles.push(tile);
        }
        tile_map.push(tiles);
    }
    tile_map = pad_tiles(tile_map);
    return tile_map;
}

function find_starting_point(tile_map: Tile[][]): Point {
    for (let y = 0; y < tile_map.length; y++) {
        for (let x = 0; x < tile_map[y].length; x++) {
            if (tile_map[y][x].symbol == "S") {
                return new Point(x, y);
            }
        }
    }
    throw "No starting point found";
}

function get_connected_directions(point: Point, tile_map: Tile[][]): string {
    let code = ["0", "0", "0", "0"];
    for (let dir of Directions) {
        let neighbour = point.get_neighbour_in_direction(dir);
        if (tile_map[neighbour.y][neighbour.x].connections.get(opposite(dir))) {
            code[dir] = "1";
        }
    }
    return code.join("");
}
function replace_starting_point(tile_map: Tile[][], point: Point) {
    tile_map[point.y][point.x] = new Tile(CODE_SYMBOLS[get_connected_directions(point, tile_map)]);
}

function follow_path(tile_map: Tile[][], point: Point): Set<string> {
    let current_point = point
    let current_direction = tile_map[point.y][point.x].connections.keys().next().value;
    let path: Set<string> = new Set<string>();
    path.add(current_point.string());
    while (true) {
        current_point = current_point.get_neighbour_in_direction(current_direction);
        current_direction = tile_map[current_point.y][current_point.x].path.get(opposite(current_direction));
        path.add(current_point.string());
        if (current_point.equals(point)) {
            break;
        }
    }
    return path
}

function scan_rows(tile_map: Tile[][], path: Set<string>): number {
    let inside_count = 0;
    for (let y = 1; y < tile_map.length - 1; y++) {
        let x = 1
        let is_inside = false;
        while (x < tile_map[y].length - 1) {
            if (!path.has(new Point(x, y).string())) {
                if (is_inside) {
                    inside_count += 1;
                }
            } else if (tile_map[y][x].symbol == "|") {
                is_inside = !is_inside;
            } else if (tile_map[y][x].symbol == "F") {
                x++;
                while (tile_map[y][x].symbol == "-") {
                    x++;
                }
                if (tile_map[y][x].symbol == "J") {
                    is_inside = !is_inside;
                }
            } else if (tile_map[y][x].symbol == "L") {
                x++;
                while (tile_map[y][x].symbol == "-") {
                    x++;
                }
                if (tile_map[y][x].symbol == "7") {
                    is_inside = !is_inside;
                }
            }
            x++;
        }
    }
    return inside_count
}

function solution_part_1() {
    let tile_map = load_data("input.txt");
    let starting_point = find_starting_point(tile_map);
    replace_starting_point(tile_map, starting_point);
    let path = follow_path(tile_map, starting_point);
    console.log(path.size / 2);
}

function solution_part_2() {
    let tile_map = load_data("input.txt");
    let starting_point = find_starting_point(tile_map);
    replace_starting_point(tile_map, starting_point);
    let path = follow_path(tile_map, starting_point);
    let inside_count = scan_rows(tile_map, path);
    console.log(inside_count);
}

solution_part_1()
solution_part_2()